### PR TITLE
Add global configuration

### LINKED ISSUE #5011: GPU memory monitor fails to log allocation.
When setting `verbosity` to 3 the memory logger should log all allocation and deallocation on GPU.  But I encountered the deallocatation error when input is cudf, saying that the pointer in SpasePage::data is never allocated, which is obviously not true.

This is a reminder for myself.  Also @RAMitchell might have some inputs here.

``` python
def test_cudf():
    X = np.random.randn(kRows, kCols)
    y = np.random.randn(kRows)

    df = pd.DataFrame(X)

    df = cudf.from_pandas(df)

    dtrain = xgb.DMatrix(df, label=y)
    bst = xgb.train({'tree_method': 'gpu_hist', 'verbosity': 2},
                    dtrain=dtrain,
                    evals=[(dtrain, 'X')],
                    num_boost_round=2)

    print('Pickle')
    import pickle
    with open('model.bst', 'bw') as fd:
        pickle.dump(bst, fd)

    with open('model.bst', 'br') as fd:
        bst: xgb.Booster = pickle.load(fd)
        dtest = xgb.DMatrix(df, label=y)
        bst.set_param({'gpu_id': '0'})
        bst.predict(dtest)
```

### PR DESCRIPTION
**Pull request by hcho3**
- [x] Add management functions for global configuration: `XGBSetGlobalConfig()`, `XGBGetGlobalConfig()`.
- [x] Add Python interface: `set_config()`, `get_config()`, and `config_context()`.
- [x] Add unit tests for Python
- [x] Add R interface: `xgb.set.config()`, `xgb.get.config()`
- [x] Add unit tests for R

Example (Python):
```python
import xgboost as xgb

# Suppress warning caused by model generated with XGBoost version < 1.0.0
with xgb.config_context(verbosity=0):
    bst = xgb.Booster(model_file='./old_model.bin')
```

Example (R):
```R
library(xgboost)

# Set verbosity level to silent (0)
xgb.set.config(verbosity = 0)
# Now global verbosity level is 0
config <- xgb.get.config()
print(config$verbosity)
# Set verbosity level to warning (1)
xgb.set.config(verbosity = 1)
# Now global verbosity level is 1
config <- xgb.get.config()
print(config$verbosity)
```

Closes #6363
Closes #5011

We can add more parameters in a future PR.

### CODE REVIEW

File: python-package/xgboost/__init__.py
Comment (trivialfis) [inline python-package/xgboost/__init__.py:34]: Is it necessary to have explicit getter and setter?

File: python-package/xgboost/config.py
Comment (trivialfis) [inline python-package/xgboost/config.py:73]: The parameter is `**new_config`, not `verbosity`.

File: python-package/xgboost/config.py
Comment (trivialfis) [inline python-package/xgboost/config.py:8]: Can we avoid listing parameter and use opaque strings similar to `booster.set_param`.

File: src/c_api/c_api.cc
Comment (trivialfis) [inline src/c_api/c_api.cc:50]: Another option is use JSON.  We have `booster.load/save_config`, maybe we can align with it.

File: src/global_config.cc
Comment (trivialfis) [inline src/global_config.cc:15]: I think we don't need to check for str equality.  The `XGBoostParameter` struct can return all unknown parameters and consume all known parameters automatically.

File: src/logging.cc
Comment (trivialfis) [inline src/logging.cc:61]: Ah, you are already using it.  Can we place this outside of logger?

**Comment by trivialfis** [comment]
Some early review comments, for discussion.

File: python-package/xgboost/__init__.py
Comment (hcho3) [reply python-package/xgboost/__init__.py:34]: I was following sklearn's config manager: https://scikit-learn.org/stable/modules/generated/sklearn.config_context.html

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:73]: This is similar to sklearn's config manager: https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/_config.py#L85

File: src/global_config.cc
Comment (hcho3) [reply src/global_config.cc:15]: Do you want me to create a new `XGBoostParameter` struct? I didn't do that because I thought you didn't like duplicating parameters. If I create `GlobalConfigurationParams` struct, it will duplicate `verbosity` parameter.

File: src/logging.cc
Comment (hcho3) [reply src/logging.cc:61]: Actually I will need to remove this check from `ConsoleLogger::Configure()`, as it's failing lots of tests.

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:8]: It's a bit nicer to explicitly parameters, so that the Python API doc will show it. Otherwise, users won't know which parameters are in the global scope. See https://scikit-learn.org/stable/modules/generated/sklearn.config_context.html

File: src/c_api/c_api.cc
Comment (hcho3) [reply src/c_api/c_api.cc:50]: Didn't think of that. Yes, that would reduce the boilerplace a bit.

File: src/global_config.cc
Comment (hcho3) [reply src/global_config.cc:15]: Actually duplicating parameters may be unavoidable, since we are already doing that in the Python layer. Let me go ahead and create `GlobalConfigurationParams`.

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:8]: Another option is to list global parameters in https://xgboost.readthedocs.io/en/latest/parameter.html.

File: python-package/xgboost/__init__.py
Comment (hcho3) [reply python-package/xgboost/__init__.py:34]: Also, the context manager is implemented in terms of the getter and the setter.

File: src/global_config.cc
Comment (trivialfis) [reply src/global_config.cc:15]: Please don't duplicate parameters.  We can have a deeper look into it later.

Comment (hcho3) [reply src/global_config.cc:15]: Got it. I won't create `GlobalConfigurationParams`

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:8]: I'm going to honor the spirit of avoiding duplication, and will avoid listing parameter in the Python later. Instead, we'll have a documented list in `parameter.rst`.

File: src/c_api/c_api.cc
Comment (hcho3) [reply src/c_api/c_api.cc:50]: @trivialfis One issue is that R does not have a built-in JSON parser, so we'd need to add a new dependency on the R package to add global config to the R package. On the other hand, the current approach does not require a new dependency.

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py]: See the latest commit, which uses opaque strings like `set_param`.

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:83]: Updated it to `**new_config`, as the parameters are no longer explicitly listed in Python.

**Comment by trivialfis** [comment]
Is it necessary for this to be blocking?  I thought it's going into the next release?

**Comment by hcho3** [comment]
@trivialfis This PR does not contain `use_rmm` or `gpu_id`. Were we not planning to address #6363 in this release?

**Comment by trivialfis** [comment]
I usually take some time for API changes (addition counts).  Some comments from @RAMitchell would also be nice.  My opinion would be we branch out whatsoever, and back port this PR if it can be merged before rc2 or final release if there's no rc2.  This way we don't have to rush.

File: src/c_api/c_api.cc
Comment (hcho3) [reply src/c_api/c_api.cc:50]: Implementing `get_config` in R turns out to be a headache, since I have to write R extension code that returns a dictionary. For the ease of maintenance, I propose that we make `jsonlite` a required dependency of XGBoost, so that `get_config` can simply return a JSON string.

**Comment by hcho3** [comment]
Got it. This PR won't block RC1.

**Comment by hcho3** [comment]
I added the R interface.

**Comment by hcho3** [comment]
@meztez This is replacement for #6363. Can you review?

File: R-package/tests/testthat/test_config.R
Comment (jameslamb) [inline R-package/tests/testthat/test_config.R:3]: this isn't necessary, and I think you can remove it. `testthat` will already handle loading `{xgboost}`, based on the commands you have in https://github.com/dmlc/xgboost/blob/master/R-package/tests/testthat.R

File: R-package/NAMESPACE
Comment (jameslamb) [inline R-package/NAMESPACE:56]: Since you're adding a new dependency, `{jsonlite}`, I expect it to show up in the diff on NAMESPACE.

I think you need to add `importFrom`s to this list: https://github.com/dmlc/xgboost/blob/master/R-package/R/xgboost.R#L78

```r
#' @importFrom jsonlite fromJSON
#' @importFrom toJSON
```

As a side note, I have a personal preference for putting these `importFrom` comments on the code that uses the imports, not grouping them all together in one file. Makes it harder to make mistakes like this, and it means that when code is removed from your package, the imports it carried with it are automatically removed too, without you needing to remember to go update a separate file. But that's bigger than this PR.

**Review by jameslamb — COMMENTED** [review]
> @jameslamb I'm quite new to the R ecosystem. Does the new addition to the R package look reasonable to you?

Looks pretty good! I left a few suggestions that I think should be addressed, though.


You should feel comfortable adding `{jsonlite}` as a dependency. It doesn't introduce any other required dependencies (https://cran.r-project.org/web/packages/jsonlite/index.html) and is used by a lot of other packages, so it is battled-tested. `{lightgbm}` depends on it for JSON parsing: https://github.com/microsoft/LightGBM/blob/637ff80038e946b3ad7520a49dbdc01ac0ae2c2d/R-package/DESCRIPTION#L51

File: R-package/NAMESPACE
Comment (hcho3) [reply R-package/NAMESPACE:56]: @jameslamb I read that `@importFrom` is not necessary when `::` operator is used to access individual functions: https://kbroman.org/pkg_primer/pages/depends.html

Comment (jameslamb) [reply R-package/NAMESPACE:56]: It's something that R people argue about haha.

I personally like to use both. I think of it like this:

* `::` = guarantees the correctness of your code (and makes it slightly faster because R doesn't search for symbols across all attached namespaces)
* `importFrom` = documentation that makes the `NAMESPACE` file readable as a source of truth for *"what are all the packages this package depends on, and which specific objects does it need?"*

File: R-package/DESCRIPTION
Comment (jameslamb) [inline R-package/DESCRIPTION:66]: @hcho3 one more thing I forgot to comment on...do you know for sure that the code you've introduced is incompatible with older versions of `{jsonlite}`? Version 1.6 is almost 2 years old so this floor is reasonable, but if you don't have a specific reason for adding this floor I'd change this to `(>= 1.0)`

I think that it's better for libraries to only have floors / ceilings on major versions (like `>=1.0`) unless they know about a specific incompatibility with some minor version. That makes your package work for the largest number of possible users.

File: R-package/NAMESPACE
Comment (hcho3) [reply R-package/NAMESPACE:56]: Got it. Let me add `importForm` for jsonlite functions.

File: R-package/DESCRIPTION
Comment (hcho3) [reply R-package/DESCRIPTION]: Given that LightGBM is using 1.0+, it should be fine for XGBoost to use it too.

File: R-package/tests/testthat/test_config.R
Comment (hcho3) [reply R-package/tests/testthat/test_config.R]: Removed in my latest commit.

**Review by jameslamb — APPROVED** [review]
R changes look good to me!

**Comment by hcho3** [comment]
@RAMitchell Can I get a review from you too?

**Review by trivialfis — COMMENTED** [review]
Could you please help taking a brief look into using it with dask?  See what's the expected out come.  For example, users will set this value on client process, then what happens on worker process?

**Comment by hcho3** [comment]
@trivialfis 

> users will set this value on client process, then what happens on worker process?

The worker process probably won't receive the new `verbosity` value. There are two options:
* Users should set `verbosity` in the Booster parameter (as before).
* Or users should use `client.run` to call `xgb.set_config` in every worker process.

**Comment by trivialfis** [comment]
Maybe we can get some inspirations from https://docs.dask.org/en/latest/configuration.html#downstream-libraries to merge xgboost's configuration into dask in client process then obtain it on worker process.  WDYT?

**Comment by trivialfis** [comment]
But whichever path we choose, I think it's best we can have a concise description over the expected behavior before merging.

**Comment by hcho3** [comment]
I'm not sure whether all global configuration should always be passed to the workers. For example we certainly do not want to pass `gpu_id` to the workers, when we do add `gpu_id` to global configuration.

If we were to decide on automatic propagation of some global parameters (like `verbosity`), we should clearly document which parameters will be propagated.

**Comment by trivialfis** [comment]
@hcho3 We certainly don't want user to set `gpu_id` on dask interface at all, not even before this PR.   I think in long term we are going to remove those parameters in training and in favor of the new ones defined here.

**Comment by hcho3** [comment]
Interesting, so you are planning to do away with `gpu_id` entirely then? In that case, we can define a precise mechanism for propagating global config from the client to workers.

**Comment by trivialfis** [comment]
@hcho3 I wanted to do that a while ago and vaguely talked about it in https://github.com/dmlc/xgboost/issues/4600 .

**Comment by trivialfis** [comment]
I'm not sure about the "right" design with dask interface.  But hopefully we can make it simple as possible.

**Comment by hcho3** [comment]
@trivialfis I documented the expected behavior of `set_config()` in a Dask cluster.

**Comment by trivialfis** [comment]
I will look deeper on Monday.  ;-)

**Comment by trivialfis** [comment]
Still WIP.  I'm trying to use `dmlc::Parameter` as a base, but it accepts only strings, which makes it a bit difficult to work with.  Will port a routine from our sklearn wrapper to config.

**Comment by hcho3** [comment]
This PR no longer blocks 1.3.0 release. Let's get this feature right, so that we can use it to solve #6297.

File: doc/parameter.rst
Comment (trivialfis) [inline doc/parameter.rst:24]: Should we deprecate the one used in training?

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:65]: I kept getting error saying that `header` was an unbound local variable.

File: python-package/xgboost/config.py
Comment (hcho3) [reply python-package/xgboost/config.py:123]: No, it should be `[str, Any]`.

File: python-package/xgboost/dask.py
Comment (hcho3) [inline python-package/xgboost/dask.py:650]: Does Dask automatically marshal the object `_global_config` to the worker processes? That's pretty cool.

**Review by hcho3 — COMMENTED** [review]
Thanks for tackling the integration with Dask. It looks good to me.

File: python-package/xgboost/dask.py
Comment (hcho3) [inline python-package/xgboost/dask.py:642]: Is it okay that `get_config()` is not async, whereas `_get_rabit_args()` is?

File: doc/parameter.rst
Comment (hcho3) [reply doc/parameter.rst:24]: Yes. How soon should we do so?

File: python-package/xgboost/dask.py
Comment (trivialfis) [reply python-package/xgboost/dask.py:642]: Get rabit args has a client running in it.

File: python-package/xgboost/dask.py
Comment (trivialfis) [reply python-package/xgboost/dask.py:650]: Cool most of the time.  Not cool when you start worrying about something got pickled. ;-)

File: doc/parameter.rst
Comment (trivialfis) [reply doc/parameter.rst:24]: Not sure.  We can discuss it for the next release when stuffs like rmm and gpu_id are ported here.

File: python-package/xgboost/config.py
Comment (trivialfis) [reply python-package/xgboost/config.py:65]: That's weird.

Comment (hcho3) [reply python-package/xgboost/config.py:65]: You can try removing `config_doc.header = header` and see if you get the same error as I did.

Comment (trivialfis) [reply python-package/xgboost/config.py:65]: @hcho3 Fixed in latest commit.

Comment (hcho3) [reply python-package/xgboost/config.py]: Great!

File: python-package/xgboost/config.py
Comment (trivialfis) [reply python-package/xgboost/config.py]: Fixed.

**Comment by trivialfis** [comment]
Added check for unknown parameter.

File: R-package/R/xgb.config.R
Comment (trivialfis) [inline R-package/R/xgb.config.R:2]: Can we use `stable` instead of `latest`?

File: R-package/R/xgb.config.R
Comment (trivialfis) [inline R-package/R/xgb.config.R:6]: Ditto.

File: R-package/R/xgb.config.R
Comment (hcho3) [reply R-package/R/xgb.config.R:2]: Sure. Go ahead.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**
